sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetime.scala

```diff
@@ -0,0 +1,255 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import java.sql.Date
+import java.text.SimpleDateFormat
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions.codegen.{GeneratedExpressionCode, CodeGenContext}
+import org.apache.spark.sql.catalyst.util.DateTimeUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+abstract class DateFormatExpression extends UnaryExpression with ExpectsInputTypes {
+  self: Product =>
+
+  protected val format: String
+
+  override def expectedChildTypes: Seq[DataType] = Seq(TimestampType)
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = child.eval(input)
+    if (valueLeft == null) {
+      null
+    } else {
+      if (format == null) {
+        null
+      } else {
+        val sdf = new SimpleDateFormat(format)
+        UTF8String.fromString(sdf.format(new Date(valueLeft.asInstanceOf[Long] / 10000)))
+      }
+    }
+  }
+
+  override protected def defineCodeGen(
+      ctx: CodeGenContext,
+      ev: GeneratedExpressionCode,
+      f: String => String): String = {
+
+    val sdf = classOf[SimpleDateFormat].getName
+    super.defineCodeGen(ctx, ev, (x) => {
+      f(s"""${ctx.stringType}.fromString((new $sdf("$format"))
+            .format(new java.sql.Date($x / 10000)))""")
+    })
+  }
+
+}
+
+case class DateFormatClass(left: Expression, right: Expression) extends BinaryExpression
+    with ExpectsInputTypes {
+
+  override def dataType: DataType = StringType
+
+  override def toString: String = s"DateFormat($left, $right)"
+
+  override def expectedChildTypes: Seq[DataType] = Seq(TimestampType, StringType)
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = left.eval(input)
+    if (valueLeft == null) {
+      null
+    } else {
+      val valueRight = right.eval(input)
+      if (valueRight == null) {
+        null
+      } else {
+        val sdf = new SimpleDateFormat(valueRight.toString)
+        UTF8String.fromString(sdf.format(new Date(valueLeft.asInstanceOf[Long] / 10000)))
+      }
+    }
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val sdf = classOf[SimpleDateFormat].getName
+    defineCodeGen(ctx, ev, (x, y) => {
+      s"""${ctx.stringType}.fromString((new $sdf($y.toString()))
+          .format(new java.sql.Date($x / 10000)))"""
+    })
+  }
+}
+
+case class Year(child: Expression) extends DateFormatExpression with ExpectsInputTypes {
+
+  override protected val format: String = "y"
+
+  override def dataType: DataType = IntegerType
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    defineCodeGen(ctx, ev, c => s"Integer.parseInt($c.toString())")
```

Comment:
Since @adrian-wang also work on datetime functions, can you two work together to avoid duplicates?